Python/traceback.c

```diff
@@ -512,8 +516,143 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent, i
     return err;
 }
 
+/* AST based Traceback Specialization
+ *
+ * When displaying a new traceback line, for certain syntactical constructs
+ * (e.g a subscript, an arithmetic operation) we try to create a representation
+ * that separates the primary source of error from the rest.
+ * 
+ * Example specialization of BinOp nodes:
+ *  Traceback (most recent call last):
+ *    File "/home/isidentical/cpython/cpython/t.py", line 10, in <module>
+ *      add_values(1, 2, 'x', 3, 4)
+ *      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ *    File "/home/isidentical/cpython/cpython/t.py", line 2, in add_values
+ *      return a + b + c + d + e
+ *             ~~~~~~^~~
+ *  TypeError: 'NoneType' object is not subscriptable
+ */
+
+static int
+extract_anchors_from_expr(PyObject *segment, expr_ty expr, int *left_anchor, int *right_anchor)
+{
+    switch (expr->kind) {
+        case BinOp_kind: {
+            PyObject *operator = PyUnicode_Substring(segment, expr->v.BinOp.left->end_col_offset,
+                                                     expr->v.BinOp.right->col_offset);
+            if (!operator) {
+                return -1;
+            }
+
+            const char *operator_str = PyUnicode_AsUTF8(operator);
+            if (!operator_str) {
+                Py_DECREF(operator);
+                return -1;
+            }
+
+            Py_ssize_t i, len = PyUnicode_GET_LENGTH(operator);
+            for (i = 0; i <= len; i++) {
+                if (operator_str[i] == ' '  ||
+                    operator_str[i] == '\t' ||
+                    operator_str[i] == '\f')
+                {
+                    continue;
+                }
+                *left_anchor = expr->v.BinOp.left->end_col_offset + i;
+                *right_anchor = expr->v.BinOp.left->end_col_offset + i + 1;
```

Comment:
This will be wrong for binary operators with more than 2 characters, e.g `//`

Reply:
Ah, makes sense. Guess we need to also either hard-code a check for the next character, or just create another while loop to find the range for `right_anchor.`

Reply:
Or track the operator in the AST? 😅

Reply:
`op` nodes don't have position information (they are even cached at Python level due to this).
https://github.com/python/cpython/blob/91a8f8c16ca9a7e2466a8241d9b41769ef97d094/Parser/Python.asdl#L99-L101

Reply:
Instead of extracting the position information, which may be quite limiting, I think a better design is make the visitor print the carets. This way you can handle assymetric information requirements in every kind of node and you don't need to contorty everything so the caller prints the thing

Reply:
> This way you can handle assymetric information requirements in every kind of node

Are there any specializations that you can think of which would benefit from this approach? The advantage of the current way is that we are just setting 2 variables in the visitors, and unifying all the printing / deciding logic in the `tb_displayline`. 

The current case is not a serious problem, since in either way we just need to do something like this; `if (i + 1<= len && !IS_WHITESPACE(operator_str[i + 1])) { ++*right_anchor; }` (if we assume all operators are either 1 or 2 characters, which I seriously think wouldn't get broken in the near future but if we want to make it more structurally sound we can always use a while loop.

Reply:
No currently, but I could imagine some cases like displaying function arguments with different carets or something. I find it more general and less coupled to the current state of things and is more "future proof". On the other hand of you think "let's go with the simpler design for now" I will not discuss that ;)

Reply:
> No currently, but I could imagine some cases like displaying function arguments with different carets or something.

Can't we not just use `call.args[0].col_offset, ..., call.keywords[-1].end_col_offset` to find the block (or do you intend to use different carets, even then we can just pass a pointer to it). The reason that I am a bit against the approach of printing on the visitor is that, it would be really hard to manage and definitely complicate the code. But if you feel strong about it, I can draft a patch about how it would look (perhaps in the next week or so, I am blocked atm).

Reply:
> But if you feel strong about i

These days I do not feel strong about *almost* anything :)